File: csharp/TruthMarkClient.cs
using System;
using System.IO;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text.Json;
using System.Threading.Tasks;

namespace TruthMark.SDK
{
    public class TruthMarkConfig
    {
        public string BaseUrl { get; set; } = "http://localhost:8000";
        public string ApiKey { get; set; }
        public int TimeoutSeconds { get; set; } = 30;
    }

    public class EncodeResult
    {
        public string Status { get; set; }
        public Metadata MetaData { get; set; }
        public string DownloadUrl { get; set; }

        public class Metadata
        {
            public double Psnr { get; set; }
            public int BitsEmbedded { get; set; }
        }
    }

    public class DecodeResult
    {
        public bool Found { get; set; }
        public string Message { get; set; }
        public double Confidence { get; set; }
    }

    public class TruthMarkClient : IDisposable
    {
        private readonly HttpClient _httpClient;
        private readonly string _baseUrl;
        private readonly string _apiKey;

        public TruthMarkClient() : this(new TruthMarkConfig()) { }

        public TruthMarkClient(TruthMarkConfig config)
        {
            _baseUrl = config.BaseUrl;
            _apiKey = config.ApiKey;
            _httpClient = new HttpClient
            {
                Timeout = TimeSpan.FromSeconds(config.TimeoutSeconds)
            };
        }

        /// <summary>
        /// Embed an invisible watermark into an image
        /// </summary>
        public async Task<EncodeResult> EncodeAsync(string imagePath, string message)
        {
            if (!File.Exists(imagePath))
            {
                throw new FileNotFoundException($"Image file not found: {imagePath}");
            }

            using var form = new MultipartFormDataContent();
            using var fileStream = File.OpenRead(imagePath);
            using var fileContent = new StreamContent(fileStream);
            
            fileContent.Headers.ContentType = MediaTypeHeaderValue.Parse("image/png");
            form.Add(fileContent, "file", Path.GetFileName(imagePath));
            form.Add(new StringContent(message), "message");

            var request = new HttpRequestMessage(HttpMethod.Post, $"{_baseUrl}/v1/encode")
            {
                Content = form
            };

            if (!string.IsNullOrEmpty(_apiKey))
            {
                request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _apiKey);
            }

            var response = await _httpClient.SendAsync(request);
            response.EnsureSuccessStatusCode();

            var responseBody = await response.Content.ReadAsStringAsync();
            return JsonSerializer.Deserialize<EncodeResult>(responseBody);
        }

        /// <summary>
        /// Extract watermark from an image
        /// </summary>
        public async Task<DecodeResult> DecodeAsync(string imagePath)
        {
            if (!File.Exists(imagePath))
            {
                throw new FileNotFoundException($"Image file not found: {imagePath}");
            }

            using var form = new MultipartFormDataContent();
            using var fileStream = File.OpenRead(imagePath);
            using var fileContent = new StreamContent(fileStream);
            
            fileContent.Headers.ContentType = MediaTypeHeaderValue.Parse("image/png");
            form.Add(fileContent, "file", Path.GetFileName(imagePath));

            var request = new HttpRequestMessage(HttpMethod.Post, $"{_baseUrl}/v1/decode")
            {
                Content = form
            };

            if (!string.IsNullOrEmpty(_apiKey))
            {
                request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _apiKey);
            }

            var response = await _httpClient.SendAsync(request);
            response.EnsureSuccessStatusCode();

            var responseBody = await response.Content.ReadAsStringAsync();
            return JsonSerializer.Deserialize<DecodeResult>(responseBody);
        }

        public void Dispose()
        {
            _httpClient?.Dispose();
        }
    }
}

File: swift/TruthMarkClient.swift
import Foundation

public struct TruthMarkConfig {
    public var baseURL: String
    public var apiKey: String?
    public var timeout: TimeInterval
    
    public init(
        baseURL: String = "http://localhost:8000",
        apiKey: String? = nil,
        timeout: TimeInterval = 30
    ) {
        self.baseURL = baseURL
        self.apiKey = apiKey
        self.timeout = timeout
    }
}

public struct EncodeResult: Codable {
    public let status: String
    public let metadata: Metadata
    public let downloadUrl: String
    
    public struct Metadata: Codable {
        public let psnr: Double
        public let bitsEmbedded: Int
        
        enum CodingKeys: String, CodingKey {
            case psnr
            case bitsEmbedded = "bits_embedded"
        }
    }
    
    enum CodingKeys: String, CodingKey {
        case status, metadata
        case downloadUrl = "download_url"
    }
}

public struct DecodeResult: Codable {
    public let found: Bool
    public let message: String?
    public let confidence: Double
}

public enum TruthMarkError: Error {
    case fileNotFound
    case networkError(Error)
    case apiError(Int, String)
    case decodingError
}

public class TruthMarkClient {
    private let config: TruthMarkConfig
    private let session: URLSession
    
    public init(config: TruthMarkConfig = TruthMarkConfig()) {
        self.config = config
        
        let configuration = URLSessionConfiguration.default
        configuration.timeoutIntervalForRequest = config.timeout
        self.session = URLSession(configuration: configuration)
    }
    
    /// Embed an invisible watermark into an image
    public func encode(imagePath: String, message: String) async throws -> EncodeResult {
        guard FileManager.default.fileExists(atPath: imagePath) else {
            throw TruthMarkError.fileNotFound
        }
        
        guard let url = URL(string: "\(config.baseURL)/v1/encode") else {
            throw TruthMarkError.apiError(0, "Invalid URL")
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        
        if let apiKey = config.apiKey {
            request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        }
        
        let boundary = UUID().uuidString
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        
        let imageData = try Data(contentsOf: URL(fileURLWithPath: imagePath))
        let fileName = URL(fileURLWithPath: imagePath).lastPathComponent
        
        var body = Data()
        
        // Add file
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"file\"; filename=\"\(fileName)\"\r\n".data(using: .utf8)!)
        body.append("Content-Type: image/png\r\n\r\n".data(using: .utf8)!)
        body.append(imageData)
        body.append("\r\n".data(using: .utf8)!)
        
        // Add message
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"message\"\r\n\r\n".data(using: .utf8)!)
        body.append(message.data(using: .utf8)!)
        body.append("\r\n".data(using: .utf8)!)
        body.append("--\(boundary)--\r\n".data(using: .utf8)!)
        
        request.httpBody = body
        
        do {
            let (data, response) = try await session.data(for: request)
            
            guard let httpResponse = response as? HTTPURLResponse else {
                throw TruthMarkError.networkError(NSError(domain: "Invalid response", code: 0))
            }
            
            guard httpResponse.statusCode == 200 else {
                throw TruthMarkError.apiError(httpResponse.statusCode, "API Error")
            }
            
            let decoder = JSONDecoder()
            return try decoder.decode(EncodeResult.self, from: data)
        } catch let error as TruthMarkError {
            throw error
        } catch {
            throw TruthMarkError.networkError(error)
        }
    }
    
    /// Extract watermark from an image
    public func decode(imagePath: String) async throws -> DecodeResult {
        guard FileManager.default.fileExists(atPath: imagePath) else {
            throw TruthMarkError.fileNotFound
        }
        
        guard let url = URL(string: "\(config.baseURL)/v1/decode") else {
            throw TruthMarkError.apiError(0, "Invalid URL")
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        
        if let apiKey = config.apiKey {
            request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        }
        
        let boundary = UUID().uuidString
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        
        let imageData = try Data(contentsOf: URL(fileURLWithPath: imagePath))
        let fileName = URL(fileURLWithPath: imagePath).lastPathComponent
        
        var body = Data()
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"file\"; filename=\"\(fileName)\"\r\n".data(using: .utf8)!)
        body.append("Content-Type: image/png\r\n\r\n".data(using: .utf8)!)
        body.append(imageData)
        body.append("\r\n".data(using: .utf8)!)
        body.append("--\(boundary)--\r\n".data(using: .utf8)!)
        
        request.httpBody = body
        
        do {
            let (data, response) = try await session.data(for: request)
            
            guard let httpResponse = response as? HTTPURLResponse else {
                throw TruthMarkError.networkError(NSError(domain: "Invalid response", code: 0))
            }
            
            guard httpResponse.statusCode == 200 else {
                throw TruthMarkError.apiError(httpResponse.statusCode, "API Error")
            }
            
            let decoder = JSONDecoder()
            return try decoder.decode(DecodeResult.self, from: data)
        } catch let error as TruthMarkError {
            throw error
        } catch {
            throw TruthMarkError.networkError(error)
        }
    }
}
